stubs/python-decouple/decouple.pyi

```diff
@@ -0,0 +1,168 @@
+TRUE_VALUES: set[str]  # undocumented
+FALSE_VALUES: set[str]  # undocumented
+
+def read_config(parser: ConfigParser, file: TextIO) -> None: ...  # undocumented
+def strtobool(value: str | bool) -> bool: ...  # undocumented
+
+class UndefinedValueError(Exception): ...
+class Undefined: ...
+
+undefined: Undefined
+
+class Config:
+    repository: _Repository
+    def __init__(self, repository: _Repository) -> None: ...
+    @overload
+    def get(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def get(self, option: str, default: _T) -> str | _T: ...
+    @overload
+    def get(self, option: str, *, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def get(self, option: str, default: str | _T, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, default: _T) -> str | _T: ...
+    @overload
+    def __call__(self, option: str, *, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str | _T, cast: Callable[[str], _T]) -> _T: ...
+
+class _Repository(Protocol):  # undocumented
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEmpty:  # undocumented
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryIni(RepositoryEmpty):
+    SECTION: str
+    parser: ConfigParser
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEnv(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositorySecret(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class AutoConfig:
+    SUPPORTED: OrderedDict[str, RepositoryEmpty]
+    encoding: str
+    search_path: str | None
+    config: Config | None
+    def __init__(self, search_path: str | None = None) -> None: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, default: _T) -> str | _T: ...
+    @overload
+    def __call__(self, option: str, *, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str | _T, cast: Callable[[str], _T]) -> _T: ...
+
+config: AutoConfig
+
+class Csv(Generic[_T, _TCsv]):
+    cast: Callable[[str], _T]
+    delimiter: str
+    strip: str
+    post_process: Callable[[Iterable[Any]], _TCsv]
```

Comment:
Shouldn't the Any be _T as the post-processor gets passed whatever came out of `cast`?

Reply:
Yes, that was my first suggestion, and we tried that, however mypy will then reject generic classes as a `post_process` since it can't seem to solve for both type vars simultaneously i.e. it will reject `post_process=set` but it would accept `post_process=set[str]`.

I agree that this should work, but it's also a very challenging scenario for an overload solver, so I am not surprised mypy can't do it yet.

Reply:
Should I add a comment that it should be `_T` but mypy wasn't able to support it.  So in the future, it's more obvious about the decision?